src/nlp/arrow_dataset.py

```diff
@@ -557,3 +558,43 @@ def apply_function_on_filtered_inputs(inputs, indices):
                 return Dataset.from_buffer(buf_writer.getvalue())
         else:
             return self
+
+    def filter(
+        self,
+        function,
+        with_indices=False,
+        **kwargs
+    ):
+        def map_function(batch, *args):
+            # flatten dict to list
+            flattened_batch = flatten_nested(batch)
+            num_items = len(batch.keys())
+            num_examples = len(flattened_batch) // num_items
+            result = defaultdict(list)
+
+            # create single examples
+            for example_pos in range(num_examples):
+                example = {}
+                for i, key in enumerate(batch.keys()):
+                    example[key] = flattened_batch[i * num_examples + example_pos]
```

Comment:
It's probably not going to work for nested structures

Reply:
Works for `trivia_qa` as well :-) It essentially just merges the value lists of all values in a `dict`

Reply:
Ok I see but if we have a batch of size two like this one:
```python
{
   "qa": {
        "question": ["foo", "bar],
        "answer": ["1", "2"]
    }
}
```
Then the reconstructed examples will be different as `flatten_nested` will return something like `["foo", "bar", "1", "2"]` and you only reconstruct the first depth of the structure (`len(batch.keys())`=1)